File: frontend/src/api/ocs.ts
import { apiClient } from "@/api/client";
import type { Paginated } from "@/types";

// OCS Inventory NG 整合。路徑帶 /api/v1 前綴（baseURL 為 /）。
// 與其他整合最大不同：帳密選用（OCS REST 預設無驗證）。

export interface OcsServer {
  id: string;
  name: string;
  source_type: string;
  base_url: string | null;
  enabled: boolean;
  verify_tls: boolean;
  api_username: string | null;
  /** 有沒有設密碼（後端不回明文） */
  has_password: boolean;
  sync_interval_seconds: number;
  stale_after_days: number;
  sync_networks: boolean;
  sync_bios: boolean;
  /** 預設關：軟體讓每台 ~2 KB → ~80 KB */
  sync_software: boolean;
  scope_subnet_ids?: string[] | null;
  detected_version: string | null;
  last_sync_at: string | null;
  last_success_at: string | null;
  last_error: string | null;
  last_cost: Record<string, unknown> | null;
}

export interface OcsWrite {
  name: string;
  base_url: string;
  enabled?: boolean;
  verify_tls?: boolean;
  api_username?: string | null;
  api_password?: string;
  /** 明確清掉已存帳密（改回無驗證） */
  clear_credentials?: boolean;
  sync_interval_seconds?: number;
  stale_after_days?: number;
  sync_networks?: boolean;
  sync_bios?: boolean;
  sync_software?: boolean;
  scope_subnet_ids?: string[];
}

export interface OcsDiagnosis {
  base_url: string | null;
  source_type: string;
  reachable?: boolean;
  computer_count?: number | null;
  incremental?: boolean;
  auth_required?: boolean;
  /** true = 這套 OCS 沒帶憑證也讀得到 → 該警告站台 */
  unauthenticated_access?: boolean;
}

export async function listOcs(): Promise<Paginated<OcsServer>> {
  const { data } = await apiClient.get<Paginated<OcsServer>>("/api/v1/ocs", {
    params: { page: 1, page_size: 200 },
  });
  return data;
}

export async function createOcs(p: OcsWrite): Promise<OcsServer> {
  const { data } = await apiClient.post<OcsServer>("/api/v1/ocs", p);
  return data;
}

export async function updateOcs(id: string, p: Partial<OcsWrite>): Promise<OcsServer> {
  const { data } = await apiClient.patch<OcsServer>(`/api/v1/ocs/${id}`, p);
  return data;
}

export async function deleteOcs(id: string): Promise<void> {
  await apiClient.delete(`/api/v1/ocs/${id}`);
}

export async function testOcs(id: string): Promise<OcsDiagnosis> {
  const { data } = await apiClient.post<OcsDiagnosis>(`/api/v1/ocs/${id}/test`);
  return data;
}

export async function syncOcs(id: string): Promise<{ task_id: string }> {
  const { data } = await apiClient.post(`/api/v1/ocs/${id}/sync`);
  return data;
}

/** OCS 盤點到的電腦，一台一筆（一台電腦的多個 IP 彙整在 ips）。 */
export interface OcsAgent {
  ocs_id: number | null;
  name: string | null;
  ips: string[];
  ip_address_ids: string[];
  os: string | null;
  agent_version: string | null;
  tag: string | null;
  last_inventory: string | null;
}

/** 有主機名稱、卻從來沒被 OCS 盤點過的 IP（同 Wazuh 的 MissingAgent）。 */
export interface OcsMissingAgent {
  ip_address_id: string;
  ip: string | null;
  hostname: string | null;
  // 所屬範圍（依區段／子網路／單位篩選用）
  subnet_id?: string | null;
  subnet_cidr?: string | null;
  section_id?: string | null;
  section_name?: string | null;
  customer_id?: string | null;
  customer_name?: string | null;
}

export async function listOcsAgents(): Promise<{ items: OcsAgent[]; total: number }> {
  const { data } = await apiClient.get("/api/v1/ocs/agents");
  return data;
}

export async function listOcsMissingAgents(): Promise<OcsMissingAgent[]> {
  const { data } = await apiClient.get<OcsMissingAgent[]>("/api/v1/ocs/missing-agents");
  return data;
}

File: backend/app/schemas/ocs.py
"""OCS Inventory NG 整合 schemas。

與其他整合最大的不同：**帳密是選用的**（OCS REST 預設無驗證），所以 `api_username` /
`api_password` 都可為 None。Phase 1 只收 REST 來源。
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Annotated, Any

from pydantic import Field, HttpUrl

from app.schemas.base import StrictModel


class OcsBase(StrictModel):
    name: Annotated[str, Field(min_length=1, max_length=128)]
    base_url: HttpUrl
    enabled: bool = True
    verify_tls: bool = True
    #: 帳密選用 —— OCS REST 預設無驗證
    api_username: Annotated[str | None, Field(max_length=128)] = None
    #: 2.11 增量跑得動每小時；2.10 只能全量，管理員應自己調長
    sync_interval_seconds: Annotated[int, Field(ge=300, le=86400)] = 3600
    stale_after_days: Annotated[int, Field(ge=1, le=3650)] = 30
    sync_networks: bool = True
    sync_bios: bool = True
    #: ⚠️ 預設關：軟體讓每台 ~2 KB → ~80 KB（5000 台 ≈ 400 MB/輪）。第一版尚未落地儲存。
    sync_software: bool = False


class OcsCreate(OcsBase):
    api_password: Annotated[str | None, Field(max_length=512)] = None


class OcsUpdate(StrictModel):
    name: Annotated[str | None, Field(min_length=1, max_length=128)] = None
    base_url: HttpUrl | None = None
    enabled: bool | None = None
    verify_tls: bool | None = None
    api_username: Annotated[str | None, Field(max_length=128)] = None
    api_password: Annotated[str | None, Field(max_length=512)] = None
    #: 傳 true 明確清掉已存的帳密（改回無驗證）
    clear_credentials: bool = False
    sync_interval_seconds: Annotated[int | None, Field(ge=300, le=86400)] = None
    stale_after_days: Annotated[int | None, Field(ge=1, le=3650)] = None
    sync_networks: bool | None = None
    sync_bios: bool | None = None
    sync_software: bool | None = None


class OcsRead(StrictModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    name: str
    source_type: str
    base_url: str | None
    enabled: bool
    verify_tls: bool
    api_username: str | None
    #: 有沒有設密碼（不回明文）
    has_password: bool = False
    sync_interval_seconds: int
    stale_after_days: int
    sync_networks: bool
    sync_bios: bool
    sync_software: bool
    detected_version: str | None
    last_sync_at: datetime | None
    last_success_at: datetime | None
    last_error: str | None
    last_cost: dict[str, Any] | None

    @classmethod
    def from_row(cls, row: Any) -> OcsRead:
        obj = cls.model_validate(row)
        obj.has_password = row.api_password_enc is not None
        return obj
